### PR TITLE
Add modebar interactions to plotly

### LINKED ISSUE #244: Add modebar interactions to plotly
There already exists the modebar marker in core and plotly for scatterplot and barchart, it should also be extended to the other visualization types. Additionally, it can be improved to look like in the sketch prototype which has icons included (they need to be added as .svg files to the static assets folder)

Sketch:

Image

Currently in visAhoi:

Image

### PR DESCRIPTION
**Pull request by dvsilviyageo**
closes #244 

Summary: 
* Added modebar interaction messages to plotly demos.
* It is generated only for the icons present.

screenshot:
image

### CODE REVIEW

File: packages/core/src/bar-chart.ts [hunk 1]
```diff
@@ -42,6 +43,25 @@ function generateMessages (
...
+    for(let i=0; i<modebar.length; i++){
+      modebarText.push(modebar.item(i)?.dataset?.title)
+    }    
+  }
+
+  let modeIconDescription = ''
+  const cameraIcon = `${modebarText.includes('Download plot as a png') ? `${SvgIcons.CAMERA} <b>Screenshot</b>: You can download a .png of the bar chart.<br/>`: ''}`
+  const zoomIcon = `${modebarText.includes('Zoom') ? `${SvgIcons.ZOOM} <b>Zooming</b>: With the left click you can zoom in the bar chart to get a more detailed view on the data.<br/>`: ''}`
+  const panIcon = `${modebarText.includes('Pan') ? `${SvgIcons.PAN} <b>Panning</b>: You can move the view left and right while dragging the mouse.<br/>`: ''}`
+  const selectionIcon = `${modebarText.includes('Box Select') ? `${SvgIcons.BOX_SELECTION} <b>Selection</b>: Drag the mouse over the bar chart to select a certain data.<br/>`: ''}`
```
Comment (dvdanielamoitzi) [inline packages/core/src/bar-chart.ts:59]:
```suggestion
  const selectionIcon = `${modebarText.includes('Box Select') ? `${SvgIcons.BOX_SELECTION} <b>Selection</b>: Drag the mouse over the bar chart to select some data.<br/>`: ''}`
```

File: packages/core/src/bar-chart.ts [hunk 2]
```diff
@@ -42,6 +43,25 @@ function generateMessages (
...
+ 
+  if(modebar){
+    for(let i=0; i<modebar.length; i++){
+      modebarText.push(modebar.item(i)?.dataset?.title)
+    }    
+  }
+
+  let modeIconDescription = ''
+  const cameraIcon = `${modebarText.includes('Download plot as a png') ? `${SvgIcons.CAMERA} <b>Screenshot</b>: You can download a .png of the bar chart.<br/>`: ''}`
+  const zoomIcon = `${modebarText.includes('Zoom') ? `${SvgIcons.ZOOM} <b>Zooming</b>: With the left click you can zoom in the bar chart to get a more detailed view on the data.<br/>`: ''}`
```
Comment (dvdanielamoitzi) [inline packages/core/src/bar-chart.ts:57]:
```suggestion
  const zoomIcon = `${modebarText.includes('Zoom') ? `${SvgIcons.ZOOM} <b>Zooming</b>: Click and drag the mouse over a certain part of the visualization to zoom in and get a more detailed view of the data.<br/>`: ''}`
```

File: packages/core/src/bar-chart.ts [hunk 3]
```diff
@@ -42,6 +43,25 @@ function generateMessages (
...
+  const selectionIcon = `${modebarText.includes('Box Select') ? `${SvgIcons.BOX_SELECTION} <b>Selection</b>: Drag the mouse over the bar chart to select a certain data.<br/>`: ''}`
+  const lassoSelectIcon = `${modebarText.includes('Lasso Select') ? `${SvgIcons.LASSO_SELECTION} <b>Lasso Select</b>: Select the desired data by drawing a lasso loop.<br/>`: ''}`
+  const zoomInIcon = `${modebarText.includes('Zoom in') ? `${SvgIcons.ZOOM_IN} <b>Zoom in</b>: With this you can zoom in the bars of the chart.<br/>`: ''}`
+  const zoomOutIcon = `${modebarText.includes('Zoom out') ? `${SvgIcons.ZOOM_OUT} <b>Zoom out:</b> With this you can zoom out the bars in the chart.<br/>`: ''}`
+  const autoScaleIcon = `${modebarText.includes('Autoscale') ? `${SvgIcons.AUTO_SCALE} <b>Autoscale</b>: Looks at all the data points in the plot and changes the layout to show all of themLooks at all the data points in the plot and changes the layout to show all of them.<br/>`: ''}`
+  const resetIcon = `${modebarText.includes('Reset axes') ? `${SvgIcons.RESET} <b>Reset</b>: It takes the chart to the inital layout settings.<br/>`: ''}`
```
Comment (dvdanielamoitzi) [inline packages/core/src/bar-chart.ts:64]:
```suggestion
  const resetIcon = `${modebarText.includes('Reset axes') ? `${SvgIcons.RESET} <b>Reset</b>: Resets the chart to the inital settings.<br/>`: ''}`
```

File: packages/core/src/bar-chart.ts [hunk 4]
```diff
@@ -42,6 +43,25 @@ function generateMessages (
...
+      modebarText.push(modebar.item(i)?.dataset?.title)
+    }    
+  }
+
+  let modeIconDescription = ''
+  const cameraIcon = `${modebarText.includes('Download plot as a png') ? `${SvgIcons.CAMERA} <b>Screenshot</b>: You can download a .png of the bar chart.<br/>`: ''}`
+  const zoomIcon = `${modebarText.includes('Zoom') ? `${SvgIcons.ZOOM} <b>Zooming</b>: With the left click you can zoom in the bar chart to get a more detailed view on the data.<br/>`: ''}`
+  const panIcon = `${modebarText.includes('Pan') ? `${SvgIcons.PAN} <b>Panning</b>: You can move the view left and right while dragging the mouse.<br/>`: ''}`
+  const selectionIcon = `${modebarText.includes('Box Select') ? `${SvgIcons.BOX_SELECTION} <b>Selection</b>: Drag the mouse over the bar chart to select a certain data.<br/>`: ''}`
+  const lassoSelectIcon = `${modebarText.includes('Lasso Select') ? `${SvgIcons.LASSO_SELECTION} <b>Lasso Select</b>: Select the desired data by drawing a lasso loop.<br/>`: ''}`
```
Comment (dvdanielamoitzi) [inline packages/core/src/bar-chart.ts:60]:
```suggestion
  const lassoSelectIcon = `${modebarText.includes('Lasso Select') ? `${SvgIcons.LASSO_SELECTION} <b>Lasso Select</b>: Select a certain subset of the data by drawing a lasso around it.<br/>`: ''}`
```

File: packages/core/src/change-matrix.ts [hunk 5]
```diff
@@ -39,6 +41,26 @@ function generateMessages (
...
+  const panIcon = `${modebarText.includes('Pan') ? `${SvgIcons.PAN} <b>Panning</b>: You can move the view left and right while dragging the mouse.<br/>`: ''}`
+  const selectionIcon = `${modebarText.includes('Box Select') ? `${SvgIcons.BOX_SELECTION} <b>Selection</b>: Drag the mouse over the chart to select a certain subset of the data.<br/>`: ''}`
+  const lassoSelectIcon = `${modebarText.includes('Lasso Select') ? `${SvgIcons.LASSO_SELECTION} <b>Lasso select</b>: Select by drawing a lasso loop in the chart.<br/>`: ''}`
+  const zoomInIcon = `${modebarText.includes('Zoom in') ? `${SvgIcons.ZOOM_IN} <b>Zoom in</b>: With this you can zoom in chart.<br/>`: ''}`
+  const zoomOutIcon = `${modebarText.includes('Zoom out') ? `${SvgIcons.ZOOM_OUT} <b>Zoom out:</b> With this you can zoom out the chart.<br/>`: ''}`
+  const autoScaleIcon = `${modebarText.includes('Autoscale') ? `${SvgIcons.AUTO_SCALE} <b>Autoscale</b>: Changes the layout to show the complete chart.<br/>`: ''}`
+  const resetIcon = `${modebarText.includes('Reset axes') ? `${SvgIcons.RESET} <b>Reset</b>: It takes the chart to the inital layout settings.<br/>`: ''}`
```
Comment (dvdanielamoitzi) [inline packages/core/src/change-matrix.ts:62]: Please replace by the general messages for all vis types

File: packages/core/src/heatmap.ts [hunk 6]
```diff
@@ -41,6 +45,27 @@ function generateMessages (
...
+  const panIcon = `${modebarText.includes('Pan') ? `${SvgIcons.PAN} <b>Panning</b>: You can move the view left and right while dragging the mouse.<br/>`: ''}`
+  const selectionIcon = `${modebarText.includes('Box Select') ? `${SvgIcons.BOX_SELECTION} <b>Selection</b>: Drag the mouse over heatmap to select a certain subset of the data.<br/>`: ''}`
+  const lassoSelectIcon = `${modebarText.includes('Lasso Select') ? `${SvgIcons.LASSO_SELECTION} <b>Lasso select</b>: Select by drawing a lasso loop.<br/>`: ''}`
+  const zoomInIcon = `${modebarText.includes('Zoom in') ? `${SvgIcons.ZOOM_IN} <b>Zoom in</b>: With this you can zoom in the heatmap.<br/>`: ''}`
+  const zoomOutIcon = `${modebarText.includes('Zoom out') ? `${SvgIcons.ZOOM_OUT} <b>Zoom out:</b> With this you can zoom out the heatmap.<br/>`: ''}`
+  const autoScaleIcon = `${modebarText.includes('Autoscale') ? `${SvgIcons.AUTO_SCALE} <b>Autoscale</b>: Changes the layout to show the complete heatmap.<br/>`: ''}`
+  const resetIcon = `${modebarText.includes('Reset axes') ? `${SvgIcons.RESET} <b>Reset</b>: It takes the heatmap to the inital layout settings.<br/>`: ''}`
```
Comment (dvdanielamoitzi) [inline packages/core/src/heatmap.ts:66]: Please replace by the general messages for all vis types

File: packages/core/src/horizon-graph.ts [hunk 7]
```diff
@@ -38,6 +40,27 @@ function generateMessages (
...
+  const panIcon = `${modebarText.includes('Pan') ? `${SvgIcons.PAN} <b>Panning</b>: You can move the view left and right while dragging the mouse.<br/>`: ''}`
+  const selectionIcon = `${modebarText.includes('Box Select') ? `${SvgIcons.BOX_SELECTION} <b>Selection</b>: Drag the mouse over the chart to select a certain subset of the data.<br/>`: ''}`
+  const lassoSelectIcon = `${modebarText.includes('Lasso Select') ? `${SvgIcons.LASSO_SELECTION} <b>Lasso select</b>: Select by drawing a lasso loop in the graph.<br/>`: ''}`
+  const zoomInIcon = `${modebarText.includes('Zoom in') ? `${SvgIcons.ZOOM_IN} <b>Zoom in</b>: With this you can zoom in graph.<br/>`: ''}`
+  const zoomOutIcon = `${modebarText.includes('Zoom out') ? `${SvgIcons.ZOOM_OUT} <b>Zoom out:</b> With this you can zoom out the graph.<br/>`: ''}`
+  const autoScaleIcon = `${modebarText.includes('Autoscale') ? `${SvgIcons.AUTO_SCALE} <b>Autoscale</b>: Changes the layout to show complete graph.<br/>`: ''}`
+  const resetIcon = `${modebarText.includes('Reset axes') ? `${SvgIcons.RESET} <b>Reset</b>: It takes the graph to the inital layout settings.<br/>`: ''}`
```
Comment (dvdanielamoitzi) [inline packages/core/src/horizon-graph.ts:62]: Please replace by the general messages for all vis types

File: packages/core/src/scatterplot.ts [hunk 8]
```diff
@@ -48,6 +52,28 @@ function generateMessages (
...
+  const selectionIcon = `${modebarText.includes('Box Select') ? `${SvgIcons.BOX_SELECTION} <b>Selection</b>: Drag the mouse over the dots in the scatterplot to select a certain subset of the data.<br/>`: ''}`
+  const lassoSelectIcon = `${modebarText.includes('Lasso Select') ? `${SvgIcons.LASSO_SELECTION} <b>Lasso select</b>: Select points by drawing a lasso loop around the points in the graph.<br/>`: ''}`
+  const zoomInIcon = `${modebarText.includes('Zoom in') ? `${SvgIcons.ZOOM_IN} <b>Zoom in</b>: With this you can zoom in the data points of the graph.<br/>`: ''}`
+  const zoomOutIcon = `${modebarText.includes('Zoom out') ? `${SvgIcons.ZOOM_OUT} <b>Zoom out:</b> With this you can zoom out the data points in the graph.<br/>`: ''}`
+  const autoScaleIcon = `${modebarText.includes('Autoscale') ? `${SvgIcons.AUTO_SCALE} <b>Autoscale</b>: Looks at all the data points in the plot and changes the layout to show all of them.<br/>`: ''}`
+  const resetIcon = `${modebarText.includes('Reset axes') ? `${SvgIcons.RESET} <b>Reset</b>: It takes the plot to the inital layout settings.<br/>`: ''}`
```
Comment (dvdanielamoitzi) [inline packages/core/src/scatterplot.ts:73]: Please replace by the general messages for all vis types

File: packages/core/src/treemap.ts [hunk 9]
```diff
@@ -35,8 +37,28 @@ function generateMessages (
...
+  const panIcon = `${modebarText.includes('Pan') ? `${SvgIcons.PAN} <b>Panning</b>: You can move the view left and right while dragging the mouse.<br/`: ''}`
+  const selectionIcon = `${modebarText.includes('Box Select') ? `${SvgIcons.BOX_SELECTION} <b>Selection</b>: Drag the mouse over to select a certain subset of the data.<br/>`: ''}`
+  const lassoSelectIcon = `${modebarText.includes('Lasso Select') ? `${SvgIcons.LASSO_SELECTION} <b>Lasso select</b>: Select by drawing a lasso loop.<br/>`: ''}`
+  const zoomInIcon = `${modebarText.includes('Zoom in') ? `${SvgIcons.ZOOM_IN} <b>Zoom in</b>: With this you can zoom in the treemap.<br/>`: ''}`
+  const zoomOutIcon = `${modebarText.includes('Zoom out') ? `${SvgIcons.ZOOM_OUT} <b>Zoom out:</b> With this you can zoom out the treemap.<br/>`: ''}`
+  const autoScaleIcon = `${modebarText.includes('Autoscale') ? `${SvgIcons.AUTO_SCALE} <b>Autoscale</b>: Shows the complete treemap.<br/>`: ''}`
+  const resetIcon = `${modebarText.includes('Reset axes') ? `${SvgIcons.RESET} <b>Reset</b>: It takes the treemap to the inital layout settings.<br/>`: ''}`
```
Comment (dvdanielamoitzi) [inline packages/core/src/treemap.ts:60]: Please replace by the general messages for all vis types

**Review by dvdanielamoitzi — CHANGES_REQUESTED** [review]
Please use the messages from the barchart to create the general interaction message and re-use them in all chart types. When doing so, provide a function e.g. getGeneralChartInteractions that returns a <string, string>Map of all chart interactions. This way, we could easily override one specific property in each visualization type.

File: packages/core/src/bar-chart.ts [hunk 10]
```diff
@@ -42,6 +43,25 @@ function generateMessages (
...
+  const zoomInIcon = `${modebarText.includes('Zoom in') ? `${SvgIcons.ZOOM_IN} <b>Zoom in</b>: With this you can zoom in the bars of the chart.<br/>`: ''}`
+  const zoomOutIcon = `${modebarText.includes('Zoom out') ? `${SvgIcons.ZOOM_OUT} <b>Zoom out:</b> With this you can zoom out the bars in the chart.<br/>`: ''}`
+  const autoScaleIcon = `${modebarText.includes('Autoscale') ? `${SvgIcons.AUTO_SCALE} <b>Autoscale</b>: Looks at all the data points in the plot and changes the layout to show all of themLooks at all the data points in the plot and changes the layout to show all of them.<br/>`: ''}`
```
[the remark this answers is not in the document]
Comment (dvsilviyageo) [reply packages/core/src/bar-chart.ts]: Added a getGeneralChartInteractions function in onboarding.ts

File: packages/core/src/treemap.ts (hunk 9 above)
Comment (dvsilviyageo) [reply packages/core/src/treemap.ts]: Replaced by general messages

File: packages/core/src/scatterplot.ts (hunk 8 above)
Comment (dvsilviyageo) [reply packages/core/src/scatterplot.ts]: Replaced by general messages

File: packages/core/src/horizon-graph.ts (hunk 7 above)
Comment (dvsilviyageo) [reply packages/core/src/horizon-graph.ts]: Replaced by general messages

File: packages/core/src/heatmap.ts (hunk 6 above)
Comment (dvsilviyageo) [reply packages/core/src/heatmap.ts]: Replaced by general messages

File: packages/core/src/change-matrix.ts (hunk 5 above)
Comment (dvsilviyageo) [reply packages/core/src/change-matrix.ts]: Replaced by general messages

File: packages/core/src/bar-chart.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (dvdanielamoitzi) [reply packages/core/src/bar-chart.ts]: I moved the function to the utils.ts file

**Review by dvdanielamoitzi — APPROVED** [review]

**Pull request merged**